flight/Libraries/alarms.c

```diff
@@ -238,6 +238,143 @@ static int32_t hasSeverity(SystemAlarmsAlarmOptions severity)
+	[SYSTEMALARMS_CONFIGERROR_NONE] = {0},
+};
+
+// DONT_BUILD_IF(NELEMENTS(CONFIG_ERROR_NAMES) != SYSTEMALARMS_CONFIGERROR_NUMELEM, AlarmArrayMismatch);
+
+static const char manual_control_names[][12] = {
+	[SYSTEMALARMS_MANUALCONTROL_SETTINGS] = "MAN:SETTINGS",
+	[SYSTEMALARMS_MANUALCONTROL_NORX] = "MAN:NO RX",
+	[SYSTEMALARMS_MANUALCONTROL_ACCESSORY] = "MAN:ACC",
+	[SYSTEMALARMS_MANUALCONTROL_ALTITUDEHOLD] = "MAN:A-HOLD",
+	[SYSTEMALARMS_MANUALCONTROL_PATHFOLLOWER] = "MAN:PATH-F",
+	[SYSTEMALARMS_MANUALCONTROL_UNDEFINED] = "MAN:UNDEF",
+	[SYSTEMALARMS_MANUALCONTROL_NONE] = {0},
+};
+
+// DONT_BUILD_IF(NELEMENTS(MANUAL_CONTROL_NAMES) != SYSTEMALARMS_MANUALCONTROL_NUMELEM, AlarmArrayMismatch);
+
+static const char boot_reason_names[][16] = {
+	[SYSTEMALARMS_REBOOTCAUSE_BROWNOUT] = "BOOT:BROWNOUT",
+	[SYSTEMALARMS_REBOOTCAUSE_PINRESET] = "BOOT:PIN RESET",
+	[SYSTEMALARMS_REBOOTCAUSE_POWERONRESET] = "BOOT:PWR ON RST",
+	[SYSTEMALARMS_REBOOTCAUSE_SOFTWARERESET] = "BOOT:SW RESET",
+	[SYSTEMALARMS_REBOOTCAUSE_INDEPENDENTWATCHDOG] "BOOT:INDY WDOG",
+	[SYSTEMALARMS_REBOOTCAUSE_WINDOWWATCHDOG] = "BOOT:WIN WDOG",
+	[SYSTEMALARMS_REBOOTCAUSE_LOWPOWER] = "BOOT:LOW POWER",
+	[SYSTEMALARMS_REBOOTCAUSE_UNDEFINED] = "BOOT:UNDEFINED",
+};
+
+const char *AlarmBootReason(uint8_t reason) {
+	if (reason >= NELEMENTS(boot_reason_names)) {
+		return (const char*)&boot_reason_names[SYSTEMALARMS_REBOOTCAUSE_UNDEFINED];
+	}
+	return (const char*)&boot_reason_names[reason];
+}
+
+int32_t AlarmString(SystemAlarmsData alarm, char *buf, size_t buflen, bool blink, uint8_t *state) {
+	*state = SYSTEMALARMS_ALARM_OK;
+	buf[0] = '\0';
+	int pos = 0;
+
+	// TODO(dustin): sort the alarms by severity.  Alarm messages
+	// will get truncated.  We want the most urgent stuff to show
+	// up first, having warnings show up only if there's space.
+	for (int i = 0; i < SYSTEMALARMS_ALARM_NUMELEM; i++) {
+		if (alarm_names[i][0] != 0 &&
+		    ((alarm.Alarm[i] == SYSTEMALARMS_ALARM_WARNING) ||
+		     (alarm.Alarm[i] == SYSTEMALARMS_ALARM_ERROR) ||
+		     (alarm.Alarm[i] == SYSTEMALARMS_ALARM_CRITICAL))) {
+
+			// Returned state is the worst state.
+			if (alarm.Alarm[i] > *state) {
+				*state = alarm.Alarm[i];
+			}
+
+			char current_msg[sizeof(*config_error_names)+1] = {0};
+			switch (i) {
+			case SYSTEMALARMS_ALARM_SYSTEMCONFIGURATION:
+				strncpy(current_msg,
+					(const char*)config_error_names[alarm.ConfigError],
+					buflen);
+				buf[sizeof(*config_error_names)] = '\0';
+				break;
+			case SYSTEMALARMS_ALARM_MANUALCONTROL:
+				strncpy(current_msg,
+					(const char*)manual_control_names[alarm.ManualControl],
+					buflen);
+				buf[sizeof(*manual_control_names)] = '\0';
+				break;
+			default:
+				strncpy(current_msg, (const char*)alarm_names[i], buflen);
+				buf[sizeof(*alarm_names)] = '\0';
+			}
+
+			int this_len = strlen(current_msg);
+
+			if (pos + this_len + 2 >= buflen)
+				break;
+
+			if ((alarm.Alarm[i] != SYSTEMALARMS_ALARM_WARNING) && !blink) {
```

Comment:
I do agree.  The problem is that it needs to blink only error/critical strings *inside* the result (which may also include warnings).  I don't like that all that much.

Alternatively, this could return a collection of structured and labeled strings and let the other side manage the concatenation/blinking (since this is really Brain-specific).

Is it worth having a [x]alarmStrings that gets populated from the call instead?